Feature/cdap 6649 sys ns guard

rohitsinha54 opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-6649
- Disallow user to use dataset in system namespace through cross namespace APIs.

Note: This PR has commits from another one: https://github.com/caskdata/cdap/pull/6403.
The only commit which needs to be reviewed is: 
CDAP-6649 Guard against accessing system namespace from programs

Commit:
CDAP-6649 Guard against accessing system namespace from programs

Addressed comments

rohitsinha54 commented:
LGTM

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/NamespaceAccessException.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Thrown when a namespace cannot be accessed.
+ */
+public class NamespaceAccessException extends AlreadyExistsException {
```
It seems like if this exception is required, it should throw a FORBIDDEN. Why not just throw UnauthorizedException instead?

rohitsinha54 replied:
No this is not required, its not used anywhere. I think I accidently committed it. Removing it.

bdmogal commented on cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestManager.java:
```diff
@@ -404,6 +404,12 @@ private File createModuleJarFile(Class<?> cls) throws IOException {
   }
 
   @Override
+  public void deleteDatasetInstance(NamespaceId namespaceId, String datasetInstanceName) throws Exception {
```
Please check if it is possible to add a test without adding these extra methods to TestBase. An equivalent of this should be possible using the DatasetAdmin.drop() API.

rohitsinha54 replied:
No this is not required, its not used anywhere. I think I accidently committed it. Removing it.

rohitsinha54 replied:
I tried to use that on your suggestion but the issue with using DatasetAdmin.drop is that it does not actually deleted the dataset instance. This is because in TestBase we are binding the DatasetFramework to Remote and the DatasetAdmin which we get  is InMemory which is because all the DatasetModule is in memory since its a unit test. Hence, calling delete on the admin does not actually deleted it from the remote dataset framework where it got added.

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -911,63 +970,114 @@ public Boolean call() throws Exception {
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthMapReduce(MapReduceManager mrManager) throws Exception {
+    addDatasetInstance(Id.Namespace.SYSTEM, "keyValueTable", "table1").create();
```
how come these don't fail with unauthorized exception?

rohitsinha54 replied:
It doesn't. Should it fail ? Isn't it that every one has access to system namespace ? I am not sure about it. How does it work when you create a namespace, the meta gets written in the system ns right ?

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -990,51 +1099,105 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
+    appManager.stopAll();
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthSpark(SparkManager sparkManager) throws Exception {
```
private

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -990,51 +1099,105 @@ public void testCrossNSDatasetAccessWithAuthMapReduce() throws Exception {
+    deleteNamespace(otherNS.getNamespaceId().toId());
+  }
+
+  public void testCrossNSDatasetAccessWithAuthSpark(SparkManager sparkManager) throws Exception {
```
private

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1145,18 +1304,46 @@ public boolean apply(Privilege input) {
     DataSetManager<KeyValueTable> outTableManager = getDataset(namespaceId.toId(), datasetName);
     KeyValueTable outputTable = outTableManager.get();
-    Assert.assertFalse(outputTable.scan(null, null).hasNext());
+    CloseableIterator<KeyValue<byte[], byte[]>> scanner = outputTable.scan(null, null);
```
try...finally

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java:
```diff
@@ -230,6 +231,11 @@ public DynamicDatasetCache getDatasetCache() {
   protected <T extends Dataset> T getDataset(String namespace, String name, Map<String, String> arguments,
                                              AccessType accessType) throws DatasetInstantiationException {
+    if (NamespaceId.SYSTEM.getNamespace().equalsIgnoreCase(namespace)) {
+      throw new DatasetInstantiationException(String.format("Dataset %s cannot be instantiated from %s namespace. " +
```
Any way to get the user id in this message? Should be able to inject `AuthenticationContext`. 

`String.format("User %s cannot access dataset %s in %s namespace", authenticationContext.getPrincipal(), name, namespace)`?

rohitsinha54 replied:
Do we need user ? No one is allowed to read/write stuff in SYSTEM NS right ? So the above message sounds more like user does not have privilege rather than dataset in SYSTEM cannot be accessed regardless who the user is it. If some users can then having user will absolutely make sense. Let me know.

bdmogal replied:
good point. I guess ignore this.

bdmogal commented on cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java:
```diff
@@ -326,6 +327,12 @@ public void onTransactionCompleted(boolean jobSucceeded, @Nullable TransactionFa
                                             AccessType accessType) throws DatasetInstantiationException {
+
+      if (NamespaceId.SYSTEM.getNamespace().equalsIgnoreCase(namespace)) {
+        throw new DatasetInstantiationException(String.format("Dataset %s cannot be instantiated from %s namespace. " +
```
ditto

bdmogal commented on cdap-unit-test/src/main/java/co/cask/cdap/test/TestBase.java:
```diff
@@ -864,6 +864,17 @@ protected static void deployDatasetModule(String moduleName,
+   *
+   * @param namespaceId namespace for the dataset
+   * @param datasetInstanceName instance name
+   * @throws Exception
```
remove this if there is no explanation for it

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -887,16 +885,77 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
     SecurityRequestContext.setUserId(BOB.getName());
 
-    // Send data to stream as BOB this ensures that BOB can do stuff in auth namespace
+    // Send data to stream as BOB this ensures that BOB can write to a stream in AUTH namespace
```
nit: don't capitalize AUTH

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -887,16 +885,77 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+    appManager.stopAll();
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthFlowlet(final FlowManager flowManager) throws Exception {
```
private

bdmogal replied:
also nit: can just call this `testSystemDatasetAccessFromFlowlet`, no?

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -887,16 +885,77 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+    deleteDatasetInstance(NamespaceId.SYSTEM, "store");
+  }
+
+  public void testCrossNSDatasetAccessWithAuthFlowlet(final FlowManager flowManager) throws Exception {
```
private

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -887,16 +885,77 @@ public void testCrossNSDatasetAccessWithAuthFlowlet() throws Exception {
+    verifyDatasetIsEmpty(NamespaceId.SYSTEM, "store");
+
+    // switch to back to ALICE
+    SecurityRequestContext.setUserId(ALICE.getName());
```
why do you need to switch back to alice here? she has admin/write on NamespaceId.SYSTEM?

rohitsinha54 replied:
The idea is that all test case function should revert the user to alice if they change it to something else. Just to have a consistent  state of user from one test method call to another. Nothing to do with the call below. If you want I can move this line at the very end.

bdmogal replied:
yea, with a comment

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -911,63 +970,114 @@ public Boolean call() throws Exception {
+    appManager.stopAll();
+  }
+
+  public void testCrossNSSystemDatasetAccessWithAuthMapReduce(MapReduceManager mrManager) throws Exception {
```
private

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -911,63 +970,114 @@ public Boolean call() throws Exception {
+    verifyProgramFailure(argsForMR, mrManager);
+    verifyDatasetIsEmpty(otherNS.getNamespaceId(), "otherTable");
+
+
```
nit: remove extra newline

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -911,63 +970,114 @@ public Boolean call() throws Exception {
+    deleteNamespace(otherNS.getNamespaceId().toId());
+  }
+
+  public void testCrossNSDatasetAccessWithAuthMapReduce(MapReduceManager mrManager) throws Exception {
```
private

bdmogal commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1145,18 +1304,46 @@ public boolean apply(Privilege input) {
+      // the program status might not have stamped as failed instantly after the program finishes so wait for
+      // sometime for it to be marked as failed.
+      try {
+        Tasks.waitFor(ProgramRunStatus.FAILED, new Callable<ProgramRunStatus>() {
```
this logic should actually be in `ProgramManager.waitForFinish()`. Can you report a jira?

anwar6953 replied:
waitForFinish only waits for it to stop running; not necessarily for FAILED status.

bdmogal replied:
should be able to wait for finish with the specified state, no? `waitForFinish(int time, TimeUnit unit, ProgramRunStatus status)`. Otherwise we do these checks all over.

rohitsinha54 replied:
Opened a jira : https://issues.cask.co/browse/CDAP-7080

anwar6953 commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -868,135 +875,370 @@ public Boolean call() throws Exception {
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return flowManager.isRunning();
```
flowManager has a waitForStatus(true) method

rohitsinha54 replied:
I am aware of waitForStatus(true) but according to my understanding it waits for started (as said in documentation) not running

anwar6953 commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1058,4 +1300,50 @@ public boolean apply(Privilege input) {
+          }
+        }, 60, TimeUnit.SECONDS);
+      } catch (TimeoutException e) {
+        Assert.fail(String.format("Found program status %s while expecting it to be %s",
```
I don't think we need this.
Tasks.waitFor should throw a meaningful message itself.

anwar6953 commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1058,4 +1300,50 @@ public boolean apply(Privilege input) {
+        Tasks.waitFor(ProgramRunStatus.FAILED, new Callable<ProgramRunStatus>() {
+          @Override
+          public ProgramRunStatus call() throws Exception {
+            return runRecord.getStatus();
```
btw, a `runRecord` object will not change its value, even if you call the `getStatus` method of it over and over again.

rohitsinha54 replied:
Good catch!

anwar6953 commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -1058,4 +1300,50 @@ public boolean apply(Privilege input) {
+    programManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    // verify program history just have failures
+    List<RunRecord> history = programManager.getHistory();
```
use the getHistory method that accepts a `ProgramRunStatus`

rohitsinha54 replied:
Yes, but I want to make sure all runs are failed. If I do it that way I will have to unnecessary have to keep a count as how many failed records I expect.

rohitsinha54 commented:
@bdmogal @anwar6953 Addressed comments. Please have another look. Second commit is the addressed comments changes. 

Build: http://builds.cask.co/browse/CDAP-RUT88-1
IT: http://builds.cask.co/browse/CDAP-ITM0-20

bdmogal commented:
lgtm,  but please get @anwar6953 to review once too

rohitsinha54 commented:
Tested on Cluster.
Waiting on build and @anwar6953 to have a final look.

anwar6953 commented on cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java:
```diff
@@ -110,6 +112,8 @@
     Constants.Explore.EXPLORE_ENABLED, false,
     Constants.Security.Authorization.CACHE_ENABLED, false
   );
+  public static final EnumSet<Action> ALL_ACTIONS = EnumSet.of(Action.READ, Action.WRITE,
```
Isn't it better to use `EnumSet.allOf(Action.class)`

anwar6953 commented:
1 minor comment.
Remainder lgtm

rohitsinha54 commented:
Addressed comments. Very minimal change was done to address last comment from Ali. Built CDAP skipping tests locally. Build has passed before the last minimal change. Merging.
